File: Src/Backend/routes/sign_language.py
from datetime import datetime

from extensions import db
from flask import Blueprint, jsonify, request
from models import Leaderboard, LearningProgress, LessonLock, User
from utils import token_required

sign_lang_bp = Blueprint("sign_lang", __name__)


@sign_lang_bp.route("/progress", methods=["GET"])
@token_required
def get_progress(current_user):
    progress = LearningProgress.query.filter_by(user_id=current_user.id).all()
    return jsonify(
        [
            {
                "lesson_title": p.lesson_title,
                "category": p.category,
                "score": p.score,
                "completed_at": p.completed_at.isoformat(),
            }
            for p in progress
        ]
    )


@sign_lang_bp.route("/progress", methods=["POST"])
@token_required
def save_progress(current_user):
    data = request.get_json()
    lesson_title = data.get("lesson_title")
    category = data.get("category")
    score = data.get("score", 0)
    time_spent = data.get("time_spent", 0)  # thời gian học tính bằng giây
    accuracy = data.get("accuracy", 0.0)  # độ chính xác (%)

    # Check if exists to update or create new
    p = LearningProgress.query.filter_by(
        user_id=current_user.id, lesson_title=lesson_title
    ).first()
    if p:
        # Update existing progress
        p.score = score
        p.best_score = max(p.best_score, score)
        p.attempts += 1
        p.time_spent += time_spent
        p.accuracy = accuracy
        p.session_count += 1
        p.completed_at = datetime.utcnow()
        p.last_attempt = datetime.utcnow()
    else:
        # Create new progress record
        p = LearningProgress(
            user_id=current_user.id,
            lesson_title=lesson_title,
            category=category,
            score=score,
            best_score=score,
            attempts=1,
            time_spent=time_spent,
            accuracy=accuracy,
            session_count=1,
            completed_at=datetime.utcnow(),
            last_attempt=datetime.utcnow(),
        )
        db.session.add(p)

    db.session.commit()

    # Update leaderboard entry
    update_leaderboard(current_user.id)

    return jsonify(
        {
            "message": "Progress saved successfully",
            "lesson_title": lesson_title,
            "score": score,
            "best_score": p.best_score,
            "attempts": p.attempts,
            "session_count": p.session_count,
        }
    )


@sign_lang_bp.route("/locks", methods=["GET"])
def get_locks():
    locks = LessonLock.query.filter_by(is_locked=True).all()
    return jsonify(
        [{"target_type": l.target_type, "target_name": l.target_name} for l in locks]
    )


@sign_lang_bp.route("/locks/toggle", methods=["POST"])
@token_required
def toggle_lock(current_user):
    if current_user.role != "admin":
        return jsonify({"message": "Unauthorized"}), 403

    data = request.get_json()
    target_type = data.get("target_type")  # 'lesson' or 'category'
    target_name = data.get("target_name")

    lock = LessonLock.query.filter_by(
        target_type=target_type, target_name=target_name
    ).first()
    if lock:
        db.session.delete(lock)
        status = "unlocked"
    else:
        lock = LessonLock(
            target_type=target_type, target_name=target_name, is_locked=True
        )
        db.session.add(lock)
        status = "locked"

    db.session.commit()
    return jsonify({"message": f"Target {status} successfully", "status": status})


@sign_lang_bp.route("/leaderboard", methods=["GET"])
def get_leaderboard():
    limit = int(request.args.get("limit", 50))
    offset = int(request.args.get("offset", 0))

    # Update ranks before fetching
    update_all_ranks()

    # Get leaderboard entries with user info
    entries = (
        db.session.query(Leaderboard, User)
        .join(User)
        .filter(User.is_banned == False)
        .order_by(Leaderboard.total_score.desc())
        .offset(offset)
        .limit(limit)
        .all()
    )

    result = []
    for entry, user in entries:
        result.append(
            {
                "rank": entry.rank,
                "user_id": user.id,
                "username": user.username,
                "avatar_url": user.avatar_url,
                "total_score": entry.total_score,
                "total_lessons_completed": entry.total_lessons_completed,
                "current_streak": entry.current_streak,
                "highest_streak": entry.highest_streak,
                "average_accuracy": entry.average_accuracy,
                "total_practice_score": entry.total_practice_score,
                "last_updated": entry.last_updated.isoformat()
                if entry.last_updated
                else None,
            }
        )

    return jsonify(result)


@sign_lang_bp.route("/leaderboard/me", methods=["GET"])
@token_required
def get_my_leaderboard(current_user):
    # Update ranks before fetching
    update_all_ranks()

    entry = Leaderboard.query.filter_by(user_id=current_user.id).first()
    if not entry:
        return jsonify({"message": "No leaderboard data found"}), 404

    return jsonify(
        {
            "rank": entry.rank,
            "user_id": current_user.id,
            "username": current_user.username,
            "avatar_url": current_user.avatar_url,
            "total_score": entry.total_score,
            "total_lessons_completed": entry.total_lessons_completed,
            "current_streak": entry.current_streak,
            "highest_streak": entry.highest_streak,
            "average_accuracy": entry.average_accuracy,
            "total_practice_score": entry.total_practice_score,
            "last_updated": entry.last_updated.isoformat()
            if entry.last_updated
            else None,
        }
    )


@sign_lang_bp.route("/leaderboard/user/<int:user_id>", methods=["GET"])
@token_required
def get_user_leaderboard(current_user, user_id):
    # Update ranks before fetching
    update_all_ranks()

    entry = Leaderboard.query.filter_by(user_id=user_id).first()
    user = User.query.get(user_id)

    if not entry or not user or user.is_banned:
        return jsonify({"message": "User not found or no data"}), 404

    return jsonify(
        {
            "rank": entry.rank,
            "user_id": user.id,
            "username": user.username,
            "avatar_url": user.avatar_url,
            "total_score": entry.total_score,
            "total_lessons_completed": entry.total_lessons_completed,
            "current_streak": entry.current_streak,
            "highest_streak": entry.highest_streak,
            "average_accuracy": entry.average_accuracy,
            "total_practice_score": entry.total_practice_score,
            "last_updated": entry.last_updated.isoformat()
            if entry.last_updated
            else None,
        }
    )


# ─── Helper Functions ─────────────────────────────────────────────────────


def update_leaderboard(user_id):
    """Update or create leaderboard entry for a user"""
    user = User.query.get(user_id)
    if not user:
        return

    # Get or create leaderboard entry
    entry = Leaderboard.query.filter_by(user_id=user_id).first()
    if not entry:
        entry = Leaderboard(user_id=user_id)
        db.session.add(entry)

    # Calculate stats from learning progress
    progress_records = LearningProgress.query.filter_by(user_id=user_id).all()

    total_score = sum(p.score for p in progress_records)
    total_lessons = len(progress_records)

    # Calculate average accuracy (assuming score = accuracy * 100)
    if total_lessons > 0:
        average_accuracy = (
            (total_score / (total_lessons * 100)) * 100 if total_lessons > 0 else 0
        )
    else:
        average_accuracy = 0.0

    entry.total_score = total_score
    entry.total_lessons_completed = total_lessons
    entry.average_accuracy = average_accuracy
    entry.last_updated = datetime.utcnow()

    db.session.commit()


def update_all_ranks():
    """Update ranks for all leaderboard entries based on total_score"""
    # Get all leaderboard entries ordered by total_score desc
    entries = Leaderboard.query.order_by(Leaderboard.total_score.desc()).all()

    # Update ranks
    for i, entry in enumerate(entries, start=1):
        entry.rank = i

    db.session.commit()

File: source/backend/routes/sign_language.js
const express = require('express');
const router = express.Router();
const { authenticateToken, requireAdmin } = require('../middleware/auth');

function leaderboardEntryToDict(entry, user) {
    return {
        rank: entry.rank,
        user_id: user.id,
        username: user.username,
        avatar_url: user.avatar_url,
        total_score: entry.total_score,
        total_lessons_completed: entry.total_lessons_completed,
        current_streak: entry.current_streak,
        highest_streak: entry.highest_streak,
        average_accuracy: entry.average_accuracy,
        total_practice_score: entry.total_practice_score,
        last_updated: entry.last_updated ? entry.last_updated.toISOString() : null
    };
}

async function updateLeaderboard(prisma, userId) {
    const user = await prisma.user.findUnique({ where: { id: userId } });
    if (!user) return;

    const progressRecords = await prisma.learningProgress.findMany({ where: { user_id: userId } });

    const total_score = progressRecords.reduce((sum, p) => sum + p.score, 0);
    const total_lessons = progressRecords.length;
    const average_accuracy = total_lessons > 0 ? (total_score / (total_lessons * 100)) * 100 : 0.0;

    await prisma.leaderboard.upsert({
        where: { user_id: userId },
        update: {
            total_score,
            total_lessons_completed: total_lessons,
            average_accuracy
        },
        create: {
            user_id: userId,
            total_score,
            total_lessons_completed: total_lessons,
            average_accuracy
        }
    });
}

async function updateAllRanks(prisma) {
    const entries = await prisma.leaderboard.findMany({ orderBy: { total_score: 'desc' } });
    for (let i = 0; i < entries.length; i++) {
        await prisma.leaderboard.update({
            where: { id: entries[i].id },
            data: { rank: i + 1 }
        });
    }
}

// GET /progress
router.get('/progress', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);

    try {
        const progress = await prisma.learningProgress.findMany({ where: { user_id: userId } });
        res.status(200).json(progress.map(p => ({
            lesson_title: p.lesson_title,
            category: p.category,
            score: p.score,
            completed_at: p.completed_at.toISOString()
        })));
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /progress
router.post('/progress', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);
    const { lesson_title, category, score = 0, time_spent = 0, accuracy = 0.0 } = req.body;

    try {
        const existing = await prisma.learningProgress.findFirst({
            where: { user_id: userId, lesson_title }
        });

        let p;
        if (existing) {
            p = await prisma.learningProgress.update({
                where: { id: existing.id },
                data: {
                    score,
                    best_score: Math.max(existing.best_score, score),
                    attempts: existing.attempts + 1,
                    time_spent: existing.time_spent + time_spent,
                    accuracy,
                    session_count: existing.session_count + 1,
                    completed_at: new Date(),
                    last_attempt: new Date()
                }
            });
        } else {
            p = await prisma.learningProgress.create({
                data: {
                    user_id: userId,
                    lesson_title,
                    category,
                    score,
                    best_score: score,
                    attempts: 1,
                    time_spent,
                    accuracy,
                    session_count: 1
                }
            });
        }

        await updateLeaderboard(prisma, userId);

        res.status(200).json({
            message: "Progress saved successfully",
            lesson_title,
            score,
            best_score: p.best_score,
            attempts: p.attempts,
            session_count: p.session_count
        });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /locks
router.get('/locks', async (req, res) => {
    const prisma = req.app.get('prisma');

    try {
        const locks = await prisma.lessonLock.findMany({ where: { is_locked: true } });
        res.status(200).json(locks.map(l => ({ target_type: l.target_type, target_name: l.target_name })));
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// POST /locks/toggle
router.post('/locks/toggle', authenticateToken, requireAdmin, async (req, res) => {
    const prisma = req.app.get('prisma');
    const { target_type, target_name } = req.body;

    try {
        const lock = await prisma.lessonLock.findFirst({ where: { target_type, target_name } });

        let status;
        if (lock) {
            await prisma.lessonLock.delete({ where: { id: lock.id } });
            status = "unlocked";
        } else {
            await prisma.lessonLock.create({ data: { target_type, target_name, is_locked: true } });
            status = "locked";
        }

        res.status(200).json({ msg: `Target ${status} successfully`, status });
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /leaderboard
router.get('/leaderboard', async (req, res) => {
    const prisma = req.app.get('prisma');
    const limit = parseInt(req.query.limit) || 50;
    const offset = parseInt(req.query.offset) || 0;

    try {
        await updateAllRanks(prisma);

        const entries = await prisma.leaderboard.findMany({
            where: { user: { is_banned: false } },
            include: { user: true },
            orderBy: { total_score: 'desc' },
            skip: offset,
            take: limit
        });

        res.status(200).json(entries.map(entry => leaderboardEntryToDict(entry, entry.user)));
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /leaderboard/me
router.get('/leaderboard/me', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.user.sub);

    try {
        await updateAllRanks(prisma);

        const entry = await prisma.leaderboard.findUnique({ where: { user_id: userId } });
        if (!entry) return res.status(404).json({ msg: "No leaderboard data found" });

        const user = await prisma.user.findUnique({ where: { id: userId } });
        res.status(200).json(leaderboardEntryToDict(entry, user));
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

// GET /leaderboard/user/:id
router.get('/leaderboard/user/:id', authenticateToken, async (req, res) => {
    const prisma = req.app.get('prisma');
    const userId = parseInt(req.params.id);

    try {
        await updateAllRanks(prisma);

        const entry = await prisma.leaderboard.findUnique({ where: { user_id: userId } });
        const user = await prisma.user.findUnique({ where: { id: userId } });

        if (!entry || !user || user.is_banned) {
            return res.status(404).json({ msg: "User not found or no data" });
        }

        res.status(200).json(leaderboardEntryToDict(entry, user));
    } catch (err) {
        console.error(err);
        res.status(500).json({ msg: "Server error" });
    }
});

module.exports = router;
